### PR TITLE
Fix query collision

### PR DESCRIPTION
**Pull request by fernandojsg**
Based on the proposal by @zeddic in https://github.com/MozillaReality/ecsy/issues/215 to fix the minification issues.
- It fixes the `Utils.queryKey` function to use `Component._typeId` instead of names, so we don't rely on names anymore other than to throw error or logs.
- We assign the `_typeId` when registering the component, the proposal to use `getTypeId()` and create the id if it doesn't exist, and otherwise just return it has a perf overhead instead of accessing a variable. To solve that I forced the components to be registered before the systems, otherwise an error will be thrown.
- Added tests

### CODE REVIEW

File: src/Utils.js [hunk 1]
```diff
@@ -25,11 +25,16 @@ export function queryKey(Components) {
   var names = [];
```
Comment (robertlong) [inline src/Utils.js:25]: I'd change `names` to `ids` now that we aren't using the component name.

File: package-lock.json [hunk 2]
```diff
@@ -6520,9 +6515,9 @@
       }
     },
     "lodash": {
```
Comment (robertlong) [inline package-lock.json:6517]: Looks like a minor patch version bump, seems fine.

File: src/System.js [hunk 3]
```diff
@@ -46,7 +47,24 @@ export class System {
         if (!Components || Components.length === 0) {
           throw new Error("'components' attribute can't be empty in a query");
         }
+
+        // Detect if the components have already been registered
+        let unregisteredComponents = Components.filter(
```
Comment (zeddic) [inline src/System.js:52]: I saw some discussion on the ecsy Discord channel where people were having Systems register their own components as a way of having singleton state. If they did this in init(), I think this will become a breaking change. Could be a non-issue if this just isn't the way things should be done. 

Another idea would be to have the `for (var queryName in this.constructor.queries)` iteration stamp queries with ids if they don't have one yet.

Comment (fernandojsg) [reply src/System.js:52]: Hi @zeddic I'm not sure if you are in the ecsy discord (https://discord.gg/cFnrQ2v ) so I paste here what I was asking the community for feedback:
```

15 PM] fernandojsg: @here I'd like to get feedback from the community in one change that could be a breaking change, in this PR: https://github.com/MozillaReality/ecsy/pull/222
TL;DR we will skip using Component.name for accessing components on entities, queries and systems, instead we will generate a Component._typeId when the component is registered.
That force us to register components before the systems that will use them. I added a check if you define a query of unregistered components, it will throw an error
GitHub
Fix query collision by fernandojsg · Pull Request #222 · MozillaRea...
Based on the proposal by @zeddic in #215 to fix the minification issues.

It fixes the Utils.queryKey function to use Component._typeId instead of names, so we don&#39;t rely on names anymore o...

[9:15 PM] fernandojsg: one of the uses cases that will break is registering the components directly on the System.init()
[9:17 PM] fernandojsg: I was thinking about taking the creation of the queries out of the System.constructor and then do something like 
 
system = new System()
system.init()
system._createQueries()
[9:18 PM] fernandojsg: but that could be another breaking change as people could be accessing the queries directly on the init  too
[9:18 PM] fernandojsg: can you tell me if you are using any of these two patterns currently?
```

Both alternatives will be a breaking change anyway

File: src/Utils.js (hunk 1 above)
Comment (fernandojsg) [reply src/Utils.js]: Thanks, fixed!
